### ISSUE TITLE
Change Forced from a type thing into a Definition thing

### ISSUE THREAD

**Issue by jespercockx** [agda/agda#2809]
Currently, the Forced tag of a function argument is attached to its type, but this has all kinds of unfortunate consequences (rgrep ignoreForced for examples). Instead, Forced should be attached to a particular definition.

**Comment by andreasabel** [comment]
I quickly read these patches. LGTM.

On master, test case #203 (in Succeed) and others are broken, reporting
```
An internal error has occurred. Please report this as a bug.
Location of the error: src/full/Agda/TypeChecking/Rules/Data.hs:228
```
Could that be related to the refactoring (or the merging of the refactoring)?

**Comment by jespercockx** [comment]
Yes, I also noticed that after pushing (sorry for being too quick with assuming master would be fine). However, I have no idea how my refactoring triggers this error. Maybe @Saizan  knows what's going on? I'll look into it further.

**Comment by andreasabel** [comment]
E.g on #203 the forcing analysis returns something different:

Previously:
```
Forcing analysis
  xs = []
```
Now:
```
Forcing analysis
  xs          = [3,2,1]
  forcedArgs  = [NotForced]
```
Not sure how this would make a difference.

But if it is just a refactoring, maybe the result should be the same?

> (sorry for being too quick with assuming master would be fine).

Ah, yes, always run the testsuite...

**Comment by andreasabel** [comment]
Ah, I have an idea.

Previously, `addForcingAnnotations` would partially normalize the type (by traversing it).
Now, this does not happen any more.
Since substitution is stricter than we want, the `applySubst (strengthenS __IMPOSSIBLE__ (size fields)) tgt` can trigger an `__IMPOSSIBLE__` too eagerly in a redex that would have gone away previously.

See #2703 where I stumbled over this recently.

**Comment by jespercockx** [comment]
> E.g on #203 the forcing analysis returns something different:

The `[3,2,1]` are the parameters of the datatype, previously they were filtered out in advance but with my new implementation of the forcing analysis this was no longer needed. Since the final result is the same as before (no arguments are forced), I don't think this is the problem.

> Previously, addForcingAnnotations would partially normalize the type (by traversing it).
> Now, this does not happen any more.

There is still a `telView` in there, so I don't see how the new code normalises the type any more or less than the old code. Anyway, I tried adding a `t <- normalise t` but I still get the same error. So the problem must be yet somewhere else.

> Ah, yes, always run the testsuite...

I did run the testsuite on stable, but I was too quick assuming it would also run fine on master...

**Comment by UlfNorell** [comment]
> I did run the testsuite on stable, but I was too quick assuming it would also run fine on master...

This is what I do all the time. As long as you watch the Travis results I think this is the most efficient way to work.

**Comment by andreasabel** [comment]
> this is the most efficient way to work.

For the individual, yes, but if you work in parallel you slow down the coworkers in case you introduce an error.

**Comment by Saizan** [comment]
To be honest, doing the telView and then strengthen is a bit fishy, I guess the better option would be to change "constructs" to return the fields and target type (if needed at least).

**Comment by jespercockx** [comment]
@Saizan If you decide to change this, could you please remove the `instantiateFull` I introduced?

**Comment by andreasabel** [comment]
Yes, but I fell into the same trap with #2703.  Intuitively, one reasons with ordinary substitution when invoking `applySubst`, and it should not crash when it applying the strengthening substitution you are applying in this case.  It should treating the exception in strengthening more lazily.

**Comment by UlfNorell** [comment]
> For the individual, yes, but if you work in parallel you slow down the coworkers in case you introduce an error.

I still maintain that it's overall more efficient, since errors are reasonably rare.

**Comment by andreasabel** [comment]
Compiling and testing master takes around 30min, but you can do this in the background while starting the next patch on stable.  It is only that you cannot push master immediately.

**Comment by nad** [comment]
I don't have separate directories for the two branches, and typically I don't even compile the master branch before I push (because often the code that I work on is identical in the two branches). I don't think this has resulted in any Travis complaints so far.
